rpc/src/client.rs

```diff
@@ -1,31 +1,45 @@
 //! Tendermint RPC client
 
-use bytes::buf::ext::BufExt;
-use hyper::header;
+use tendermint::{
+    abci::{self, Transaction},
+    block::Height,
+    evidence::Evidence,
+    net, Genesis,
+};
 
-use tendermint::abci::{self, Transaction};
-use tendermint::block::Height;
-use tendermint::evidence::Evidence;
-use tendermint::net;
-use tendermint::Genesis;
-
-use crate::{endpoint::*, Error, Request, Response};
+use crate::{
+    client::{
+        subscription::SubscriptionManager,
+        transport::{http_ws::HttpWsTransport, Transport},
+    },
+    endpoint::*,
+    Error, Request, Response,
+};
 
 pub mod event_listener;
+pub mod subscription;
+pub mod transport;
+
+#[cfg(test)]
+pub mod testing;
 
 /// Tendermint RPC client.
 ///
 /// Presently supports JSONRPC via HTTP.
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct Client {
-    /// Address of the RPC server
-    address: net::Address,
+    transport: Box<dyn Transport>,
 }
 
 impl Client {
-    /// Create a new Tendermint RPC client, connecting to the given address
-    pub fn new(address: net::Address) -> Self {
-        Self { address }
+    /// Create a new Tendermint RPC client, connecting to the given address.
+    /// By default this uses the [`HttpWsTransport`] transport layer. This
+    /// transport lazily initializes subscription mechanisms when first
+    /// subscribing to events generated by a particular query.
```

Comment:
If we have a default config we should use [Default](https://doc.rust-lang.org/std/default/trait.Default.html) for that and make it possible to use the new constructor to pass a transport. This will help with the right amount of composability.